### PR TITLE
supports llama-dybatch-V1

### PR DESCRIPTION
**Pull request by carryyu**
### PR types

New features
### PR changes

Models
### Description

supports llama-dybatch-V1

### CODE REVIEW

File: paddlenlp/ops/generation/encode_rotary_qk.cu [hunk 1]
```diff
@@ -0,0 +1,225 @@
+// Copyright (c) 2023 PaddlePaddle Authors. All Rights Reserved.
```
Comment (wj-Mcat) [inline paddlenlp/ops/generation/encode_rotary_qk.cu]: `paddlenlp/ops/generation` 目录下的 ops 我这边推荐使用：`csrc`这个目录，考虑如下：

1. 为了和主流框架自定义算子风格保持一致。
2. `csrc`目录有比较强的**分类说明**：C 相关的代码（只能够是 ops）
3. 未来 paddlenlp 也可能会有更多的加速自定义算子，此时也是可以放在这个里面，而不仅仅只是针对于生成式相关。


参考项目：

* https://github.com/Dao-AILab/flash-attention
* https://github.com/vllm-project/vllm
* https://github.com/TimDettmers/bitsandbytes

以上开源项目的自定义算子都是放在 csrc 目录下，所以建议为了和主流开源项目保持一致的风格，建议采用`csrc`这个目录。

File: paddlenlp/ops/generation/setup_cuda.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from paddle.utils.cpp_extension import CUDAExtension, setup
+
+setup(
+    name="custom_setup_ops",
```
Comment (wj-Mcat) [inline paddlenlp/ops/generation/setup_cuda.py]: 这里的 name 建议改成：`paddlenlp_ops`，从语义上更适合 paddlenlp，属于 paddlenlp 的一部分。

File: paddlenlp/transformers/fused_multi_transformer_fine_grained.py [hunk 3]
```diff
@@ -0,0 +1,670 @@
...
+
+
+# for distributed tensor model parallel
+def _set_var_distributed(var):
+    if var is None:
+        return
+
+    var.is_distributed = True
+
+    if not in_dynamic_mode():
```
Comment (wj-Mcat) [inline paddlenlp/transformers/fused_multi_transformer_fine_grained.py]: 这个 branch 会执行到吗？ 这个函数只是在创建类的时候才会用到，可是创建模型是在动态图模式下，并非是静态图模式下，所以这个 branch 应该是不会走到的。

除非是在静态图模式下创建类，这种场景在 paddlenlp 中从来没有实现过。

File: paddlenlp/transformers/fused_multi_transformer_fine_grained.py [hunk 4]
```diff
@@ -0,0 +1,670 @@
...
+            shift,
+            smooth,
+            act_method,
+            compute_dtype,
+            quant_scale,
+            quant_round_type,
+            quant_max_bound,
+            quant_min_bound,
+        )
+    else:
```
Comment (wj-Mcat) [inline paddlenlp/transformers/fused_multi_transformer_fine_grained.py]: 这里的代码是否可简化成：

```python
if in_dynamic_mode():
    return ...
helper = LayerHelper("fused_bias_act")
...
```

我看你下面的部分代码也是这样写的。

File: paddlenlp/transformers/llama/modeling.py [hunk 5]
```diff
@@ -1120,6 +1131,256 @@ def forward(
...
+        # get ring_id
+        ring_id = -1
+        try:
+            hcg = fleet.get_hybrid_communicate_group()
+            model_parallel_group = hcg.get_model_parallel_group()
+            ring_id = model_parallel_group.id
+        except:
+            pass
+
+        ln_scale_attrs = [paddle.ParamAttr(name="fusellama.{}.ln_scale".format(i)) for i in range(self.num_layers)]
```
Comment (wj-Mcat) [inline paddlenlp/transformers/llama/modeling.py]: 【可暂不调整】关于 fusedmt 中的parameter name 命名，我有不同的想法：尽量保持和原生动态图保持一致。

比如说：LLama 中的 `llama.layers[0].input_layernorm.weight` 就直接对应着：`fusellama.0.ln_scale`这个参数，然后在fusedmt 中定义的时候就直接写成：

`ln_scale_attrs = [paddle.ParamAttr(name="llama.layers[0].input_layernorm.weight".format(i)) for i in range(self.num_layers)]`

这样有如下优点：

1. 在模型定义的时候，语义更加清晰。
2. 在 set_state_dict 函数中只需要针对于 qkv transpose 就行，其他参数可直接 set 进去，甚至可以通过配置不重写 set_state_dict 函数，大大减少相关代码，从而减少 fusedmt 模型适配成本。

**Review by wj-Mcat — CHANGES_REQUESTED** [review]
我觉得大部分的工作都非常棒，有几点想跟你讨论的。

另外，后面有时间可以也加一加相关单测，目前 paddlenlp 合入进去的相关东西一般都是要加的。

File: paddlenlp/ops/generation/encode_rotary_qk.cu (hunk 1 above)
Comment (carryyu) [reply paddlenlp/ops/generation/encode_rotary_qk.cu]: 收到，请问这个csrc目录放在哪个路径下

File: paddlenlp/transformers/fused_multi_transformer_fine_grained.py (hunk 3 above)
Comment (carryyu) [reply paddlenlp/transformers/fused_multi_transformer_fine_grained.py]: 支持静态图接口 是为了在 动转静导出的时候使用的

File: paddlenlp/ops/generation/setup_cuda.py (hunk 2 above)
Comment (carryyu) [reply paddlenlp/ops/generation/setup_cuda.py]: 已修改

File: paddlenlp/transformers/fused_multi_transformer_fine_grained.py (hunk 4 above)
Comment (carryyu) [reply paddlenlp/transformers/fused_multi_transformer_fine_grained.py]: 已修改

File: paddlenlp/transformers/llama/modeling.py (hunk 5 above)
Comment (carryyu) [reply paddlenlp/transformers/llama/modeling.py]: 收到，后续会调整

File: paddlenlp/transformers/fused_multi_transformer_fine_grained.py [hunk 6]
```diff
@@ -0,0 +1,669 @@
+# Copyright (c) 2023 PaddlePaddle Authors. All Rights Reserved.
```
Comment (qingqing01) [inline paddlenlp/transformers/fused_multi_transformer_fine_grained.py]: 1. fused_transformer_layers.py 
2. 另外一些API合入框架之后，这里就不需要了吧

File: llm/llama/dybatch/export_generation_model.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
+# Copyright (c) 2023 PaddlePaddle Authors. All Rights Reserved.
```
Comment (qingqing01) [inline llm/llama/dybatch/export_generation_model.py]: 都已经在llm目录里， 文件命名可以简单一些， export_model.py

File: paddlenlp/transformers/fused_multi_transformer_fine_grained.py [hunk 8]
```diff
@@ -0,0 +1,669 @@
...
+            "residual_alpha": residual_alpha,
+            "norm_type": norm_type,
+            "begin_norm_axis": begin_norm_axis,
+        },
+        outputs=outputs_dict,
+    )
+    return (out, residual_out) if residual is not None else out
+
+
+def mmha_wrapper(
```
Comment (xiaoxiaohehe001) [inline paddlenlp/transformers/fused_multi_transformer_fine_grained.py]: api 直接调用这些就可以 from paddle.incubate.nn.functional import masked_multihead_attention

**Review by qingqing01 — COMMENTED** [review]
需要补充使用文档

File: paddlenlp/ops/generation/encode_rotary_qk.cu (hunk 1 above)
Comment (wj-Mcat) [reply paddlenlp/ops/generation/encode_rotary_qk.cu]: 我觉得就想像 其他库一样放置在根目录下，你觉得呢？

File: llm/llama/dybatch/export_generation_model.py (hunk 7 above)
Comment (wj-Mcat) [reply llm/llama/dybatch/export_generation_model.py]: 我觉得是 ok 的，就直接用`export_model.py` 命名。

File: paddlenlp/ops/generation/encode_rotary_qk.cu (hunk 1 above)
Comment (carryyu) [reply paddlenlp/ops/generation/encode_rotary_qk.cu]: OK

File: paddlenlp/transformers/fused_multi_transformer_fine_grained.py (hunk 8 above)
Comment (carryyu) [reply paddlenlp/transformers/fused_multi_transformer_fine_grained.py]: 已修改

File: llm/llama/dybatch/export_generation_model.py (hunk 7 above)
Comment (carryyu) [reply llm/llama/dybatch/export_generation_model.py]: 已修改

File: paddlenlp/transformers/fused_multi_transformer_fine_grained.py (hunk 6 above)
Comment (carryyu) [reply paddlenlp/transformers/fused_multi_transformer_fine_grained.py]: 已修改

File: paddlenlp/ops/generation/encode_rotary_qk.cu (hunk 1 above)
Comment (carryyu) [reply paddlenlp/ops/generation/encode_rotary_qk.cu]: 已修改

**Comment by carryyu** [comment]
> 需要补充使用文档

已添加

File: csrc/encode_rotary_qk.cu [hunk 9]
```diff
@@ -0,0 +1,225 @@
+// Copyright (c) 2023 PaddlePaddle Authors. All Rights Reserved.
```
Comment (wawltor) [inline csrc/encode_rotary_qk.cu]: 这块的目录建议在csrc的目录下新增一个generation的子目录，便于后续的其他custom op的拓展

File: llm/llama/dybatch/README.md [hunk 10]
```diff
@@ -0,0 +1,21 @@
+# LLaMA DyBatch
```
Comment (wawltor) [inline llm/llama/dybatch/README.md]: 目前在LLM目录模型的使用方法基本得到统一，微调、预测、量化相关的脚本都是共用一套

动态插入的脚本是否可以得到统一

File: paddlenlp/transformers/fused_transformer_layers.py [hunk 11]
```diff
@@ -0,0 +1,517 @@
...
+                ln_bias = self.create_parameter(
+                    attr=ln_bias_attr, 
+                    shape=[embed_dim], 
+                    is_bias=True,
+                    dtype=self._norm_weight_dtype, 
+                )
+            
+            qkv_weight = self.create_parameter(
+                shape=[3 * num_heads * self.head_dim, embed_dim]
+                if trans_qkvw
```
Comment (xiaoxiaohehe001) [inline paddlenlp/transformers/fused_transformer_layers.py]: trans_qkvw False 修改下。

Comment (carryyu) [reply paddlenlp/transformers/fused_transformer_layers.py]: 已修改

File: llm/llama/dybatch/split_weight.py [hunk 12]
```diff
@@ -0,0 +1,105 @@
+# Copyright (c) 2023 PaddlePaddle Authors. All Rights Reserved.
```
Comment (heavengate) [inline llm/llama/dybatch/split_weight.py]: 这里可以服用PaddleNLP自己的权重切分方法，xxx.from_pretrain里会做权重切分

File: llm/llama/dybatch/utils.py [hunk 13]
```diff
@@ -0,0 +1,247 @@
+# Copyright (c) 2023 PaddlePaddle Authors. All Rights Reserved.
```
Comment (heavengate) [inline llm/llama/dybatch/utils.py]: 这个建议放到paddlenlp/transformers目录，组织方式 @wj-Mcat 帮忙看下~

File: llm/llama/dybatch/README.md (hunk 10 above)
Comment (carryyu) [reply llm/llama/dybatch/README.md]: 这部分可能不太好统一，后续还有各种量化方法，全部放到一起显得不够清晰，或者在主README里面加一下跳转链接这样呢

File: csrc/encode_rotary_qk.cu (hunk 9 above)
Comment (carryyu) [reply csrc/encode_rotary_qk.cu]: 已修改

File: llm/llama/dybatch/split_weight.py (hunk 12 above)
Comment (carryyu) [reply llm/llama/dybatch/split_weight.py]: 这个from_pretrain应该切分有问题，会炸显存，所以重新写了一套

Comment (heavengate) [reply llm/llama/dybatch/split_weight.py]: 这里需要看看有个通用的权重切分方法 @wj-Mcat @wawltor

File: llm/llama/dybatch/export_model.py [hunk 14]
```diff
@@ -0,0 +1,147 @@
+# Copyright (c) 2023 PaddlePaddle Authors. All Rights Reserved.
```
Comment (heavengate) [inline llm/llama/dybatch/export_model.py]: 这个目录下的文件看下能否删掉，在llm/llama目录的文件加一个`--enable_dybatch`的参数，通过分支来维护

**Review by heavengate — COMMENTED** [review]
确认下是否是在tests/transformer/llama目录下添加单测

File: experimental/inference/llama/README.md [hunk 15]
```diff
@@ -0,0 +1,19 @@
+# LLaMA Inference
```
Comment (heavengate) [inline experimental/inference/llama/README.md]: 1. 按llm目录的组织方式，这里.sh文件删掉吧，文档按Python命令方式给一下，区分一下单卡和多卡
2. 多卡权重拆分的统一脚本 @wj-Mcat 看下

File: experimental/inference/llama/run.sh [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+
+export FLAGS_cache_inference_while_scope=1
+export FLAGS_call_stack_level=2
+export GLOG_logtostderr=1
+export GLOG_v=0
+export FLAGS_control_flow_use_new_executor=1
+export FLAGS_new_executor_serial_run=1
+export FLAGS_allocator_strategy=naive_best_fit
+export FLAGS_fraction_of_gpu_memory_to_use=0.95
+export FLAGS_use_cutlass_fmha=1
```
Comment (heavengate) [inline experimental/inference/llama/run.sh]: 1. 非必要的flag建议删掉，如log相关
2. 其余flag在README里简单说明一下作用

File: experimental/inference/llama/README.md (hunk 15 above)
Comment (wj-Mcat) [reply experimental/inference/llama/README.md]: > 按llm目录的组织方式，这里.sh文件删掉吧，文档按Python命令方式给一下，区分一下单卡和多卡

我在最新的 commit 当中已经删掉了。

**Comment by wj-Mcat** [comment]
截止目前，完成了 InferenceModel：

- [x] 单卡动态图、动转静和静态图验证
- [x] 多卡动态图、动转静和静态图验证
- [x] REAMD 文档的调整
- [x] inferenceModel 的调整

File: paddlenlp/transformers/llama/modeling.py [hunk 17]
```diff
@@ -153,7 +153,12 @@ def parallel_matmul(x: Tensor, y: Tensor, tensor_parallel_output=True):
     except:
         is_fleet_init = False
 
-    if is_fleet_init and tensor_parallel_degree > 1 and y.is_distributed:
+    if paddle.in_dynamic_mode():
+        y_is_distributed = y.is_distributed
+    else:
+        y_is_distributed = tensor_parallel_degree > 1
```
Comment (wj-Mcat) [inline paddlenlp/transformers/llama/modeling.py:159]: 动态图下 y.is_distributed 为真实值，可是在静态图下y.is_distributed 一直为 False，于是会影响最终 Logits 的维度，从而影响解码的精度。

在此处针对于静态图做了一定的适配。

**Review by wj-Mcat — APPROVED** [review]

File: llm/predictor.py [hunk 18]
```diff
@@ -242,53 +250,296 @@ def _infer(self, inputs: dict[str, np.ndarray]):
...
+            tokens: np.ndarray = load_real_time_tokens()
+            decoded_predictions = self.tokenizer.batch_decode(
+                tokens.tolist(), skip_special_tokens=True, clean_up_tokenization_spaces=False
+            )
+            return decoded_predictions
+        else:
+            return None
+
+
+class DygraphInferencePredictor(BasePredictor):
```
Comment (wawltor) [inline llm/predictor.py:369]: 这里的命名后续可以修改一下，之前的理解是dygraph表示动态图，inference表示静态图推理


自己记个TODO吧

File: llm/predictor.py [hunk 19]
```diff
@@ -242,53 +250,296 @@ def _infer(self, inputs: dict[str, np.ndarray]):
         return decoded_ids
 
 
-def create_predictor(predictor_args: PredictorArgument, model_args: ModelArgument):
+class StaticInferencePredictor(BasePredictor):
```
Comment (wawltor) [inline llm/predictor.py:253]: 同时这里区分动态batch和非动态batch的了

Comment (wj-Mcat) [reply llm/predictor.py:253]: 这里有两个 flag：
* mode: dygraph, static
* inference_model: bool 类型

通过以上两个 flag 来控制这四种情况。

File: llm/predictor.py (hunk 18 above)
Comment (wj-Mcat) [reply llm/predictor.py:369]: 这里的命名暂时没有比较优雅合适的名字：

* DygraphInferencePredictor（中庸）
* DygraphinferenceModelPredictor（太长）
* DIPredictor（缩写，什么鬼）

大家有什么合适的名字也可以来参与讨论。

File: llm/predictor.py (hunk 19 above)
Comment (wj-Mcat) [reply llm/predictor.py:253]: dygraph -> dynamic

**Review by wawltor — APPROVED** [review]
LGTM
